galsim/utilities.py

```diff
@@ -1913,3 +1913,244 @@ def __exit__(self, type, value, traceback):
+    # elements of objs are hashable (and that they have unique hashes!, which is what we're trying
+    # to test!.  So instead, we just loop over all combinations.
+    for i, obji in enumerate(objs):
+        assert obji == obji
+        assert not (obji != obji)
+        # Could probably start the next loop at `i+1`, but we start at 0 for completeness
+        # (and to verify a != b implies b != a)
+        for j, objj in enumerate(objs):
+            if i == j:
+                continue
+            assert obji != objj, ("Found equivalent objects {0} == {1} at indices {2} and {3}"
+                                  .format(obji, objj, i, j))
+
+    if not check_hash:
+        return
+    # Now check that all hashes are unique (if the items are hashable).
+    if not isinstance(objs[0], Hashable):
+        return
+    hashes = [hash(obj) for obj in objs]
+    if not (len(hashes) == len(set(hashes))):  # pragma: no cover
+        for k, v in Counter(hashes).items():
+            if v <= 1:
+                continue
+            print("Found multiple equivalent object hashes:")
+            for i, obj in enumerate(objs):
+                if hash(obj) == k:
+                    print(i, repr(obj))
+    assert len(hashes) == len(set(hashes))
+
+
+def timer(f):
+    """A decorator that reports how long a function took to run.
+
+    In GalSim we decorate all of our tests with this to try to watch for long-running tests.
+    """
+    import functools
+
+    @functools.wraps(f)
+    def f2(*args, **kwargs):
+        import time
+        t0 = time.time()
+        result = f(*args, **kwargs)
+        t1 = time.time()
+        fname = repr(f).split()[1]
+        print('time for %s = %.2f' % (fname, t1-t0))
+        return result
+    return f2
+
+
+class CaptureLog:
+    """A context manager that saves logging output into a string that is accessible for
+    checking in unit tests.
+
+    After exiting the context, the attribute `output` will have the logging output.
+
+    Sample usage:
+
+            >>> with CaptureLog() as cl:
+            ...     cl.logger.info('Do some stuff')
+            >>> assert cl.output == 'Do some stuff'
+
+    """
+    def __init__(self, level=3):
+        from io import StringIO
+        import logging
+
+        logging_levels = { 0: logging.CRITICAL,
+                           1: logging.WARNING,
+                           2: logging.INFO,
+                           3: logging.DEBUG }
+        self.logger = logging.getLogger('CaptureLog')
+        self.logger.setLevel(logging_levels[level])
+        self.stream = StringIO()
+        self.handler = logging.StreamHandler(self.stream)
+        self.logger.addHandler(self.handler)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self.handler.flush()
+        self.output = self.stream.getvalue().strip()
+        self.handler.close()
+
+
+# Context to make it easier to profile bits of the code
+class profile:
```

Comment:
FWIW, I'd actually love to rename this one so it doesn't conflict with using `@profile` via [line_profiler](https://github.com/pyutils/line_profiler#line_profiler).

Reply:
OK.  Suggestions?  Maybe just capitalize?  `with Profile():`  That would fit in with most of our GalSim style for classes.

Reply:
I was also thinking of renaming a couple others.  `do_pickle` and `all_diff` aren't great names either, tbh.  I was thinking 'check_pickle` and `check_all_diff` for those.

Reply:
Yeah, these all sound better to me.